ipa-core/src/protocol/ipa_prf/boolean_ops/sigmoid.rs

```diff
@@ -0,0 +1,254 @@
+        a_times_b_and_not_b(&ctx, record_id, &seg_5_or_6, &bits[4], 8),
+        seg_7_or_8.multiply(&bits[4], ctx.narrow(&ThirtyTwoBitStep::from(10)), record_id),
+    )
+    .await?;
+
+    let segments = [&seg_1, &seg_2, &seg_3, &seg_4, &seg_5, &seg_6];
+
+    let (
+        bit_4_and_segments,
+        bit_5_and_segments,
+        bit_6_and_segments,
+        bit_7_and_segments,
+        (positive_outside_of_seg_1, positive_outside_of_seg_1_and_2),
+    ) = try_join5(
+        bit_segments_cross_product(&ctx, record_id, &bits[3], &segments[0..6], 11),
+        bit_segments_cross_product(&ctx, record_id, &bits[2], &segments[0..5], 17),
+        bit_segments_cross_product(&ctx, record_id, &bits[1], &segments[0..4], 22),
+        bit_segments_cross_product(&ctx, record_id, &bits[0], &segments[0..3], 26),
+        try_join(
+            seg_1.clone().not().multiply(
+                &sign_bit_not,
+                ctx.narrow(&ThirtyTwoBitStep::from(29)),
+                record_id,
+            ),
+            seg_1_or_2.not().multiply(
+                &sign_bit_not,
+                ctx.narrow(&ThirtyTwoBitStep::from(30)),
+                record_id,
+            ),
+        ),
+    )
+    .await?;
+
+    Ok(BitDecomposed::new([
+        seg_7
+            + positive_outside_of_seg_1_and_2
+            + &bit_7_and_segments[2]
+            + &bit_6_and_segments[3]
+            + &bit_5_and_segments[4]
+            + &bit_4_and_segments[5],
+        seg_6
+            + positive_outside_of_seg_1
+            + &bit_7_and_segments[1]
+            + &bit_6_and_segments[2]
+            + &bit_5_and_segments[3]
+            + &bit_4_and_segments[4],
+        seg_5
+            + &sign_bit_not
+            + &bit_7_and_segments[0]
+            + &bit_6_and_segments[1]
+            + &bit_5_and_segments[2]
+            + &bit_4_and_segments[3],
+        seg_4
+            + &sign_bit_not
+            + &bit_6_and_segments[0]
+            + &bit_5_and_segments[1]
+            + &bit_4_and_segments[2],
+        seg_3 + &sign_bit_not + &bit_5_and_segments[0] + &bit_4_and_segments[1],
+        seg_2 + &sign_bit_not + &bit_4_and_segments[0],
+        seg_1 + &sign_bit_not,
+        sign_bit_not,
+    ]))
+}
+
+#[cfg(all(test, unit_test))]
+mod test {
+    use std::iter;
+
+    use crate::{
+        ff::{boolean::Boolean, boolean_array::BA8, I128Conversions, U128Conversions},
+        protocol::{context::Context, ipa_prf::boolean_ops::sigmoid::sigmoid, RecordId},
+        secret_sharing::{replicated::semi_honest::AdditiveShare, BitDecomposed, TransposeFrom},
+        test_executor::run,
+        test_fixture::{Reconstruct, Runner, TestWorld},
+    };
+
+    fn piecewise_linear_sigmoid_approximation(x: i128) -> u128 {
+        if x < -112 {
+            return 0;
+        }
+        if x < -96 {
+            return 1;
+        }
+        if x < -80 {
+            return 2 + (u128::try_from(x + 96).unwrap() >> 3);
```

Comment:
I think there are too many `.unwrap` calls and you could benefit from changing the API of this function to 

```rust
// not sure about the error type, but it should be the same
fn piecewise_linear_sigmoid_approximation(x: i128) -> Result<u128, IntError> {
   match x {
        i128::MIN..-112 => 0,
        -112..-96 => 1,
        -96..-80 => 2 + u128::try_from(x + 96)? >> 3,
        ...
        96..112 => 254,
        _ => 255
   }
}
```

match is optional, you can still use `if` if you don't want to repeat the range boundaries at the expense of using 3 lines per statement instead of one